File: controls/src/VehicleDynamics.cc
#include "controls/VehicleDynamics.hh"

#include <iostream>
#include <math.h>
#include <chrono>

namespace MuddSub::Controls
{


// Nonlinear vehicle dynamics
VehicleDynamics::VehicleDynamics()
{

  // Wait for parameters to be loaded onto server (read from URDF)
  bool ready = false;
  while(!nh_.hasParam("dynamics_ready") && !ready)
  {
    ROS_WARN("Waiting for dynamics params to be loaded into server");
    ros::Duration(.1).sleep();
  }

  // Load parameters for vehicle dynamics
  nh_.getParam("robot_name", robotInfo_.name_);
  std::string paramRoot = "dynamics/"+robotInfo_.name_+"/";
  nh_.getParam(paramRoot+"center_of_gravity", robotInfo_.centerOfGravity_);
  nh_.getParam(paramRoot+"center_of_buoyancy", robotInfo_.centerOfBuoyancy_);
  nh_.getParam(paramRoot+"mass", robotInfo_.mass_);
  nh_.getParam(paramRoot+"buoyancy", robotInfo_.buoyancy_);
  nh_.getParam(paramRoot+"added_mass", robotInfo_.addedMass_);
  nh_.getParam(paramRoot+"linear_damping_coeffs", robotInfo_.linearDampingCoefficients_);
  nh_.getParam(paramRoot+"rate", rate_);
  nh_.getParam(paramRoot+"inertia", robotInfo_.inertia_);

  // Construct the inertia tensor, represented as 3x3 matrix (Eigen3d)
  robotInfo_.inertiaTensor_ <<
    robotInfo_.inertia_["ixx"], -1*robotInfo_.inertia_["ixy"], -1*robotInfo_.inertia_["ixz"],
    -1*robotInfo_.inertia_["ixy"], robotInfo_.inertia_["iyy"], -1*robotInfo_.inertia_["iyz"],
    -1*robotInfo_.inertia_["ixz"], -1*robotInfo_.inertia_["iyz"], robotInfo_.inertia_["izz"];

  ROS_INFO("Robot dynamics loaded into VehicleDynamics");

  robotInfo_.centerOfGravityVector_ << robotInfo_.centerOfGravity_["x"],
                                       robotInfo_.centerOfGravity_["y"],
                                       robotInfo_.centerOfGravity_["z"];

  robotInfo_.centerOfBuoyancyVector_ << robotInfo_.centerOfBuoyancy_["x"],
                                        robotInfo_.centerOfBuoyancy_["y"],
                                        robotInfo_.centerOfBuoyancy_["z"];

}

// Clone required by Control Toolbox (pure virtual) replaced with smart pointers
VehicleDynamics* VehicleDynamics::clone() const
{
  return new VehicleDynamics(*this);
}

// Clone the current instance
std::shared_ptr<VehicleDynamics> VehicleDynamics::smartClone() const
{
  return std::shared_ptr<VehicleDynamics>(clone());
}


// Given the current state and control input, find the derivative.
// Note: Since this function is called as part of the numerical differentiation
//       routine, speed is critical.
void VehicleDynamics::computeControlledDynamics(const stateVector_t &x,
                                                const ctTime_t& t,
                                                const controlVector_t &u,
                                                stateVector_t &derivative)
{
  // Segment the state vector into position, velocity, attitute, and rotational velocity.
  // Names kept brief for the sake of large equations.
  const auto& position = x.segment<3>(0);
  const auto& attitude = x.segment<3>(3);
  const auto& v1 = x.segment<3>(6);
  const auto& v2 = x.segment<3>(9);

  // Centripital and coriolis forces
  const auto& C = buildCMatrix(v1, v2);

  // Compute the force of gravity, in the frame of the robot
  const auto& Gf = buildGravityMatrix(attitude);

  // Find transformation between world frame and robot frame
  const auto& Jn = buildJnMatrix(attitude);

  // The generalized mass matrix comprises actual mass and hydrodynamic added mass
  // Todo: This doesn't need to be calculated at every iteration
  const auto& M = buildMassMatrix();
  Eigen::Matrix<double, 6, 6> mInv;

  // TODO: Make this faster!
  if(M.determinant() == 0)
    mInv = Eigen::Matrix<double, 6, 6>::Zero();
  else
    mInv = M.inverse();

  // In state space form: x' = f(x) + g(u)
  // Refer to Chin 2013, p. 138 for a full explanation.
  Eigen::Matrix<double, 12, 12> f1;
  f1 << Eigen::Matrix<double, 6, 6>::Zero(), Jn,
        Eigen::Matrix<double, 6, 6>::Zero(), mInv;

  Eigen::Matrix<double, 12, 1> f2, f, g, xT;
  xT = x.transpose();

  Eigen::Matrix<double, 6, 1> generalizedGravity = -1*(mInv*Gf);
  f2 << Eigen::Matrix<double, 6, 1>::Zero(), generalizedGravity;

  f = f1*xT + f2;
  Eigen::Matrix<double, 6, 1> controlForces = mInv*u;
  g << Eigen::Matrix<double, 6, 1>::Zero(), controlForces;
  derivative = (f+g).transpose();
}

void VehicleDynamics::computeDynamics(const stateVector_t& state, const ctTime_t& t, stateVector_t& deriv)
{
  const auto& controller = getController();
  controlVector_t controlAction;
  if(controllerSet_)
  {
    controller->computeControl(state, t, controlAction);
  }
  else
    controlAction = controlVector_t::Zero();
  computeControlledDynamics(state, t, controlAction, deriv);
}

// Coriolis and centripital forces
Eigen::Matrix<double, 6, 6> VehicleDynamics::buildCMatrix(const Eigen::Vector3d& v1, const Eigen::Vector3d& v2)
{
  Eigen::Matrix3d C12 = -1*robotInfo_.mass_*
                              (
                                S(v1) + S(v2)*S(robotInfo_.centerOfGravityVector_)
                              );

  Eigen::Matrix3d C21 = -1*C12.transpose();

  Eigen::Matrix3d C22 = -1 * S(robotInfo_.inertiaTensor_* v2);

  // Fossen 2011, p. 65
  Eigen::Matrix<double, 6, 6> Crb;
  Crb << Eigen::Matrix3d::Zero(), C12,
         C21, C22;
  Eigen::Matrix<double, 6, 6> Ca;

  // Fossen 2011, p. 121
  Ca << 0, 0, 0, 0, -1*robotInfo_.addedMass_["z"]*v1(2), robotInfo_.addedMass_["y"]*v1(1),
        0, 0, 0, robotInfo_.addedMass_["z"]*v1(2), 0, -1*robotInfo_.addedMass_["x"]*v1(0),
        0, 0, 0, -1*robotInfo_.addedMass_["y"]*v1(1), robotInfo_.addedMass_["x"]*v1(0), 0,
        0, -1*robotInfo_.addedMass_["z"]*v1(2), robotInfo_.addedMass_["y"]*v1(1), 0, -1*robotInfo_.addedMass_["r"]*v2(2), robotInfo_.addedMass_["q"]*v2(1),
        robotInfo_.addedMass_["z"]*v1(2), 0, -1*robotInfo_.addedMass_["x"]*v1(0), robotInfo_.addedMass_["r"]*v2(2), 0, -1*robotInfo_.addedMass_["p"]*v2(0),
        -1*robotInfo_.addedMass_["y"]*v1(1), robotInfo_.addedMass_["x"]*v1(0), 0, -1*robotInfo_.addedMass_["q"]*v2(1), robotInfo_.addedMass_["p"] * v2(0), 0;

  return Ca + Crb;

}

// Generalized mass matrix
Eigen::Matrix<double, 6, 6> VehicleDynamics::buildMassMatrix()
{
  Eigen::Matrix<double, 6, 6> Mrb, Ma;
  Eigen::Matrix3d Mm = I_*robotInfo_.mass_;

  Mrb << Mm, -1*robotInfo_.mass_*S(robotInfo_.centerOfGravityVector_),
         robotInfo_.mass_*S(robotInfo_.centerOfGravityVector_), robotInfo_.inertiaTensor_;

  Eigen::Matrix<double, 1, 6> addedMassVector;
  std::map<std::string, double>& addMass = robotInfo_.addedMass_;
  addedMassVector << addMass["x"], addMass["y"], addMass["z"], addMass["p"], addMass["q"], addMass["r"];
  Ma = addedMassVector.asDiagonal();
  return Ma + Mrb;
}

// skew-symmetric matrix for 3-vector cross products
Eigen::Matrix3d VehicleDynamics::S(const Eigen::Vector3d& vec) const
{
  Eigen::Matrix3d S;
  S << 0, -1*vec(2), vec(1),
       vec(2), 0, -1*vec(0),
       -1*vec(1), vec(0), 0;
  return S;
}

// Gravity in robot's frame
Eigen::Matrix<double, 6, 1> VehicleDynamics::buildGravityMatrix(const Eigen::Vector3d& attitude)
{

  // Here we re-name variables for the sake of making equations semi-readable, and in the
  // notation of Fossen, 2011, p. 60

  double cPhi, sPhi, cTheta, sTheta, tTheta, cTsi, sTsi;
  cPhi = cos(attitude(0));
  sPhi = sin(attitude(0));
  cTheta = cos(attitude(1));
  sTheta = sin(attitude(1));
  tTheta = tan(attitude(1));
  cTsi = cos(attitude(2));
  sTsi = sin(attitude(2));

  // Weight and buoyancy.
  const auto& w = gravity_*robotInfo_.mass_;
  const auto& b = gravity_*robotInfo_.buoyancy_;

  // effective weight
  const auto& ew = w-b;

  // Centers of gravity and buoyancy.
  const auto& cgX = robotInfo_.centerOfGravity_["x"];
  const auto& cgY = robotInfo_.centerOfGravity_["y"];
  const auto& cgZ = robotInfo_.centerOfGravity_["z"];
  const auto& cbX = robotInfo_.centerOfBuoyancy_["x"];
  const auto& cbY = robotInfo_.centerOfBuoyancy_["y"];
  const auto& cbZ = robotInfo_.centerOfBuoyancy_["z"];

  // Fossen 2013, p. 60
  Eigen::Matrix<double, 6, 1> result;
  result << ew*sTheta,
           -1*ew*cTheta*sPhi,
           -1*ew*cTheta*cPhi,
           -1*(cgY*w-cbY*b)*cTheta*cPhi+(cgZ*w-cbZ*b)*cTheta*sPhi,
           (cgZ*w-cbZ*b)*sTheta+(cgX*w-cbX*b)*cTheta*cPhi,
           -1*(cgX*w-cbX*b)*cTheta*sPhi-(cgY*w-cbY*b)*sTheta;

  return result;
}

Eigen::Matrix<double, 6, 6> VehicleDynamics::buildJnMatrix(const Eigen::Vector3d& attitude)
{
  // Todo: this shouldn't be copy-pasta'd
  double cPhi, sPhi, cTheta, sTheta, tTheta, cTsi, sTsi;
  cPhi = cos(attitude(0));
  sPhi = sin(attitude(0));
  cTheta = cos(attitude(1));
  sTheta = sin(attitude(1));
  tTheta = tan(attitude(1));
  cTsi = cos(attitude(2));
  sTsi = sin(attitude(2));

  Eigen::Matrix3d Rb, Tt;

  // Fossen 2011, p. 22
  Rb  << cTsi*cTheta, (-sTsi*cPhi + cTsi*sTheta*sPhi), (sTsi*sPhi+cTsi*cPhi*sTheta),
          sTsi*cTheta, (cTsi*cPhi + sPhi*sTheta*sTsi), (-cTsi*sPhi+sTheta*sTsi*cTheta),
          -sTheta, cTheta*sPhi, cTheta*cPhi;

  // Fossen 2011, p. 25
  Tt << 1, sPhi*tTheta, cPhi*tTheta,
        0, cPhi, -sPhi,
        0, (sPhi/cTheta), (cPhi/cTheta);

  Eigen::Matrix<double, 6, 6> Jn;
  // Fossen 2011, p. 26
  Jn << Rb, Eigen::Matrix3d::Zero(),
        Eigen::Matrix3d::Zero(), Tt;

  return Jn;
}

void VehicleDynamics::setController(const controllerPtr_t& controller)
{
  controllerSet_ = true;
  controller_ = controller;
}

} // Namespace MuddSub::Controls

File: controls/scripts/VehicleDynamics.py
#!/usr/bin/env python

import rospy, control
import numpy as np

class VehicleDynamics:

    def __init__(self):
        while(not rospy.has_param("dynamics_ready") and not rospy.get_param("dynamics_ready")):
            rospy.sleep(0.1)

        self.robotName = rospy.get_param("robot_name")
        dynamics = rospy.get_param("dynamics/"+self.robotName)

        inertia = dynamics["inertia"]
        self.I = {'xx': inertia['ixx'], 'xy': inertia['ixy'], 'xz': inertia['ixz'], \
                                        'yy': inertia['iyy'], 'yz': inertia['iyz'], \
                                                              'zz': inertia['izz']}

        self.inertiaTensor = np.array([[self.I['xx'], -self.I['xy'], -self.I['xz']],
                                       [-self.I['xy'], self.I['yy'], -self.I['yz']],
                                       [-self.I['xz'], -self.I['yz'], self.I['zz']]])

        addedMass = dynamics["added_mass"]
        self.addedMass = np.diag([addedMass['x'], addedMass['y'], addedMass['z'],
                                  addedMass['k'], addedMass['m'], addedMass['n']])

        cg = dynamics["center_of_gravity"]
        cb = dynamics["center_of_buoyancy"]

        self.centerOfGravity = np.array([cg['x'], cg['y'], cg['z']])
        self.centerOfBuoyancy = np.array([cb['x'], cb['y'], cb['z']])

        self.gravity = 10

        #jk lol
        self.gravity -= 0.2

        self.rate = dynamics["rate"]


        self.mass = dynamics["mass"]
        self.buoyancy = dynamics["buoyancy"]

        self.weight = self.gravity * self.mass
        self.buoyancy = self.buoyancy * self.mass
        self.effectiveWeight = self.gravity*(self.mass - self.buoyancy)

        linearDamping = dynamics["linear_damping_coeffs"]
        self.D = -1*np.diag([linearDamping['x'], linearDamping['y'], linearDamping['z'],
                             linearDamping['k'], linearDamping['m'], linearDamping['n']])

        self.x = np.zeros(12)
        self.y = np.zeros(12)
        self.u = np.zeros(6)

        self.nonlinearIoSys = control.iosys.NonlinearIOSystem(self, inputs=6, outputs=12, states=12, dt = self.rate)
        self.linearizedSystem = None

        self.Q = np.identity(12)*100
        self.R = np.identity(6)

        self.xRef = np.zeros(12)


    # Because of how linearization works, we need to explicitly pass in x and u,
    #  even though it's a class member.
    def __call__(self, t, x, u, params={}):
        xPos = x[0]
        yPos = x[1]
        zPos = x[2]
        roll = x[3]
        pitch = x[4]
        yaw = x[5]

        xVel = x[6]
        yVel = x[7]
        zVel = x[8]
        rollVel = x[9]
        pitchVel = x[10]
        yawVel = x[11]

        v1 = x[6:9]
        v2 = x[9:12]


        cgX = self.centerOfGravity[0]
        cgY = self.centerOfGravity[1]
        cgZ = self.centerOfGravity[2]

        cbX = self.centerOfBuoyancy[0]
        cbY = self.centerOfBuoyancy[1]
        cbZ = self.centerOfBuoyancy[2]

        # Coriolis and centripital forces
        # Fossen 2011, p. 65
        C11 = np.identity(3)
        C12 = -self.mass * self.S(v1) - self.mass*np.matmul(self.S(v2), self.S(self.centerOfGravity))
        C21 = -1*np.transpose(C12)
        C22 = -self.S(np.matmul(self.inertiaTensor, v2))

        CRB  = np.vstack((np.hstack((C11, C12)), np.hstack((C21, C22))))

        MaX = self.addedMass[0,0]
        MaY = self.addedMass[1,1]
        MaZ = self.addedMass[2,2]
        MaRoll = self.addedMass[3,3]
        MaPitch = self.addedMass[4,4]
        MaYaw = self.addedMass[5,5]


        # Fossen 2011, p. 121
        Ca = np.array([
                            [0, 0, 0, 0, -MaZ*zVel, MaY*yVel],
                            [0, 0, 0, MaZ*zVel, 0, -MaX*xVel],
                            [0, 0, 0, -MaY*yVel, MaX*xVel, 0],
                            [0, -MaZ*zVel, MaY*yVel, 0, -MaYaw*yawVel, MaPitch*pitchVel],
                            [MaZ*zVel, 0, -MaX*xVel, MaYaw*yawVel, 0, -MaRoll*rollVel],
                            [-MaY*yVel, MaX*xVel, 0, -MaPitch*pitchVel, MaRoll * rollVel, 0]
                          ])

        C = Ca + CRB


        # Masses
        Mm = self.mass * np.identity(3)

        np.hstack((self.mass*self.S(self.centerOfGravity), self.inertiaTensor))

        MRB = np.vstack((np.hstack((Mm, -self.mass*self.S(self.centerOfGravity))),
                        np.hstack((self.mass*self.S(self.centerOfGravity), self.inertiaTensor))))
        M = self.addedMass + MRB



        # Transformation between world and robot frame
        # OK sorry... renamed for the sake of consistency with textbook
        phi = roll
        theta = pitch
        tsi = yaw
        cPhi = np.cos(phi);
        sPhi = np.sin(phi);
        cTheta = np.cos(theta);
        sTheta = np.sin(theta);
        tTheta = np.tan(theta);
        cTsi = np.cos(tsi);
        sTsi = np.sin(tsi);

        # Fossen 2011, p. 22
        Rb = np.array([[cTsi*cTheta, (-sTsi*cPhi + cTsi*sTheta*sPhi), (sTsi*sPhi+cTsi*cPhi*sTheta)],
                        [sTsi*cTheta, (cTsi*cPhi + sPhi*sTheta*sTsi), (-cTsi*sPhi+sTheta*sTsi*cTheta)],
                        [-sTheta, cTheta*sPhi, cTheta*cPhi]])

        # Fossen 2011, p. 25
        Tt = np.array([[1, sPhi*tTheta, cPhi*tTheta],
                       [0, cPhi, -sPhi],
                       [0, (sPhi/cTheta), (cPhi/cTheta)]]);

        # Fossen 2011, p. 26
        Jn = np.vstack((np.hstack((Rb, np.zeros((3,3)))), np.hstack((np.zeros((3,3)),Tt))))

        # Put force of gravity in frame of robot
        # Fossen 2011, p. 60
        self.Gf = np.array([[self.effectiveWeight* sTheta],
                      [-self.effectiveWeight * cTheta*sPhi],
                      [-self.effectiveWeight * cTheta*cPhi],
                      [-(cgY*self.weight - cbY*self.buoyancy)*cTheta*cPhi+(cgZ*self.weight-cbZ*self.buoyancy)*cTheta*sPhi],
                      [(cgZ*self.weight-cbZ*self.buoyancy)*sTheta+(cgX*self.weight-cbX*self.buoyancy)*cTheta*cPhi],
                      [-(cgX*self.weight-cbX*self.buoyancy)*cTheta*sPhi-(cgY*self.weight-cbY*self.buoyancy)*sTheta]])


        mInv = np.zeros((6,6))
        if np.linalg.det(M) != 0:
            mInv = np.linalg.inv(M)

        #Chin 2013, p. 138
        f1 = np.vstack((np.hstack((np.zeros((6,6)), Jn)), np.hstack((np.zeros((6,6)), -np.matmul(mInv,(C+self.D))))))
        f2 = np.vstack((np.zeros((6,1)), -np.matmul(mInv,self.Gf)))
        f = np.matmul(f1,x.reshape((12,1))) + f2
        g = np.vstack((np.zeros((6,1)), np.matmul(mInv,u).reshape((6,1))))
        return f+g


    def S(self, p):

        return np.array([[0, -p[2], p[1]],
                        [p[2], 0, -p[0]],
                        [-p[1], p[0], 0]])

    def simLQR(self, duration):
        dt = 1/self.rate

        for i in range(int(self.rate*duration)):
            linearizedSystem = control.linearize(self.nonlinearIoSys, self.x, self.u)
            K,_,_ = control.lqr(linearizedSystem, self.Q, self.R)

            err = self.x - self.xRef

            for i in range(3,6):
                err[i] = min(abs(err[i]), abs(360-err[i]))*np.sign(err[i])

            self.u = np.matmul(K, err.reshape((12,1)))

            xDot = self(0, self.x, self.u)

            self.x += xDot * dt


if __name__ == '__main__':
    rospy.init_node("VehicleDynamics")
    dynamics = VehicleDynamics()

    dt = 1/dynamics.rate

    dynamics.xRef = np.hstack((np.ones(3), np.zeros(9)))

    sys = control.iosys.NonlinearIOSystem(dynamics, inputs=6, outputs=12, states=12, name="RobotDynamics")

    dynamics.simLQR(0.1)
